### PR TITLE
dts: bindings: Add memory-region property

### PR DESCRIPTION
**Pull request by awojasinski**
Since the introduction of the `zephyr,memory-region` compatible to create new
memory region, drivers can now use the newly defined region in the DT by using
the macro LINKER_DT_NODE_REGION_NAME() on the node_id to retrieve the region
name and possibly allocated data or variables in that region (or for whatever
other use). This is assuming though that the driver knows in advance the node_id to
reference.

A better approach is to add a new generic 'memory-region' property that can be
used by any driver to reference a region by using a phandle, so that the driver
can use that to retrieve the memory-region of interest.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>
Signed-off-by: Carlo Caione <ccaione@baylibre.com>

### CODE REVIEW

**Comment by awojasinski-nordicsemi** [comment]
cc @carlocaione

**Review by galak — CHANGES_REQUESTED** [review]
I'm not sure I follow how adding this new property accomplishes what you describe.  Can you explain this further.

Also, base.yaml is probably not the best location for this as that has been used for common properties that have a well established usage in devicetree outside of just Zephyr.

(Marking this 'request changes' til we decide the use and best location for this property)

**Comment by JordanYates** [comment]
I can see the use, but it seems very specific to certain drivers. Why not leave it to specific bindings to add the property?
For example, lots of devices have a reset pin, but we haven't added `reset-gpis` to `base.yml`.

**Comment by carlocaione** [comment]
> I can see the use, but it seems very specific to certain drivers. Why not leave it to specific bindings to add the property?

yeah, that is the way it was done in the original PR #45128 but IMO that would bring a lot of redundant code and copy&paste. What about adding a new yaml like `memory-user-device.yaml` to be imported?

**Comment by carlocaione** [comment]
@mbolivar-nordic @JordanYates any feedback on this?

**Comment by JordanYates** [comment]
I agree with @galak that `base.yaml` isn't the best spot for this.
Some drivers might want multiple `memory-region`'s, most want none.
Maybe a `easydma.yaml` that contains the common property and the drivers just `include` that?
The scope in the original PR was only nordic drivers, if we end up with multiple manufacturers with similar files it will be easier to work out what a common `memory-region.yaml` needs to look like.

File: dts/bindings/dma/nordic,easydma-device.yaml
Comment (mbolivar-nordic) [reply dts/bindings/dma/nordic,easydma-device.yaml:7]: I agree with @anangl that a comment would be better than a description since this is an include file, not a binding for a compatible.

I don't see any `memory-region` property defined in `nordic,nrf-spim.yaml`, so I am confused about the second comment, though.

**Review by mbolivar-nordic — COMMENTED** [review]
Agreed that base.yaml is not the right place for this. Beyond the existing arguments against it, remember that "base properties" are a separate column in each binding's page in the binding index, like this one:

https://docs.zephyrproject.org/latest/build/dts/api/bindings/spi/nordic%2Cnrf-spim.html#dtbinding-nordic-nrf-spim

Using an include file as is being done here will mean that `memory-region` will show up in the "Node specific properties" section, where it should be, instead of the "Base properties", where it's kind of hidden.

I don't understand the exact suggestion to add this to `nordic,nrf-spim` or `nordic,nrf-twim`, but I do agree it would be nice to see this new file actually included somewhere in the upstream zephyr repository before we merge this.

File: dts/bindings/dma/nordic,easydma-device.yaml
Comment (anangl) [reply dts/bindings/dma/nordic,easydma-device.yaml:7]: > I don't see any `memory-region` property defined in `nordic,nrf-spim.yaml`, so I am confused about the second comment, though.

As I understand it, that's the purpose of the added file, to be included from bindings for nRF peripherals that use EasyDMA, so that this property can be defined in their devicetree nodes if needed. That's why I think this PR should also present the use of that file.

Comment (awojasinski-nordicsemi) [reply dts/bindings/dma/nordic,easydma-device.yaml:7]: Sure I can add the inclusion of this binding here, but initially, it was done here https://github.com/zephyrproject-rtos/zephyr/pull/45128 and I thought that after this PR is merged I'll align the one I mentioned. So @mbolivar-nordic and @anangl would you like to have it done here?

Comment (mbolivar-nordic) [reply dts/bindings/dma/nordic,easydma-device.yaml:7]: >  So @mbolivar-nordic and @anangl would you like to have it done here?

Yes, I think that makes sense.

File: dts/bindings/dma/nordic,easydma-device.yaml
Comment (mbolivar-nordic) [inline dts/bindings/dma/nordic,easydma-device.yaml:4]: Typo for "featuring". Not a blocking issue for me since this is just a comment.

Comment (carlocaione) [inline dts/bindings/dma/nordic,easydma-device.yaml:4]: `featuring`

**Review by mbolivar-nordic — COMMENTED** [review]
DTS changes look good now, but I'm concerned about merging this 3 days before the feature freeze without the corresponding driver changes as well. That would create a Zephyr v3.1 where the devicetree bindings do not reflect the driver behavior. As far as I can tell, drivers are not using the memory-region property yet.

File: dts/bindings/audio/nordic,nrf-pdm.yaml
Comment (anangl) [inline dts/bindings/audio/nordic,nrf-pdm.yaml:8]: If this commit is to add such inclusions for _all_ peripherals that use EasyDMA, as its message claims, at least the following are missing: AAR, CCM, ECB, I2S, QSPI, RADIO, SAADC, USBD.

Comment (awojasinski-nordicsemi) [reply dts/bindings/audio/nordic,nrf-pdm.yaml]: Together with @mbolivar-nordic, we decided to add this property only to peripherals that utilize it in shims so currently it is supported only in UARTE and SPIM, other peripherals will be aligned in the future

**Review by mbolivar-nordic — DISMISSED** [review]
Devicetree changes look good to me and are acceptable for v3.1 since the PR now contains the accompanying driver changes. I will leave the review of the drivers themselves to @anangl .

**Comment by awojasinski-nordicsemi** [comment]
@anangl any feedback on this?

@galak do you have any objections to the current implementation?

**Comment by awojasinski-nordicsemi** [comment]
@mbolivar-nordic this PR doesn't have a very high priority. It can wait until v3.2. I just wanted to ping people to align code to their comments and have the coding part done.

File: drivers/serial/uart_nrfx_uarte.c
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:2153]: I'd suggest to get rid of `UARTE_DMA_MEMORY_REGION()` and move this definition into its place. Then it will be easier to see what it is all about at first glance, I think.
```suggestion
#define UARTE_MEMORY_SECTION(idx)					       \
	COND_CODE_1(UARTE_HAS_PROP(idx, memory_region),			       \
		(__attribute__((__section__(LINKER_DT_NODE_REGION_NAME(	       \
			DT_PHANDLE(UARTE(idx), memory_region)))))),	       \
		())
```

File: drivers/serial/uart_nrfx_uarte.c
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:2135]:
```suggestion
#define UARTE_ASYNC(idx)						       \
	IF_ENABLED(CONFIG_UART_##idx##_ASYNC, (				       \
		static uint8_t						       \
			uarte##idx##_tx_cache[CONFIG_UART_ASYNC_TX_CACHE_SIZE] \
			UARTE_MEMORY_SECTION(idx);			       \
		struct uarte_async_cb uarte##idx##_async = {		       \
			.hw_rx_counting =				       \
				IS_ENABLED(CONFIG_UART_##idx##_NRF_HW_ASYNC),  \
```

File: drivers/serial/uart_nrfx_uarte.c
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:2143]:
```suggestion
		(static uint8_t uarte##idx##_tx_buffer			       \
			[MIN(CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE,	       \
			     BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE))]      \
			UARTE_MEMORY_SECTION(idx);			       \
```

File: drivers/spi/spi_nrfx_spim.c
Comment (anangl) [inline drivers/spi/spi_nrfx_spim.c:565]:
```suggestion
		IF_ENABLED(SPI_BUFFER_IN_RAM,				       \
			(.buffer = spim_##idx##_buffer,))		       \
```

File: drivers/spi/spi_nrfx_spim.c
Comment (anangl) [inline drivers/spi/spi_nrfx_spim.c:559]:
```suggestion
	IF_ENABLED(SPI_BUFFER_IN_RAM,					       \
		(static uint8_t spim_##idx##_buffer			       \
			[CONFIG_SPI_NRFX_RAM_BUFFER_SIZE]		       \
			SPIM_MEMORY_SECTION(idx);))			       \
```

File: drivers/spi/spi_nrfx_spim.c
Comment (anangl) [inline drivers/spi/spi_nrfx_spim.c:599]: See https://github.com/zephyrproject-rtos/zephyr/pull/45142#discussion_r877872676.

**Review by anangl — DISMISSED** [review]

**Review by galak — APPROVED** [review]

File: dts/bindings/dma/nordic,easydma-device.yaml
Comment (galak) [inline dts/bindings/dma/nordic,easydma-device.yaml:6]: Rather than putting this here, can you mimic what is done in linux and add `bindings/reserved-memory/memory-region.yaml `

See: https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/reserved-memory/memory-region.yaml

**Review by galak — CHANGES_REQUESTED** [review]
In general looks fine, just want to move the property to a more generic binding location that mimics the linux binding.

File: dts/bindings/dma/nordic,easydma-device.yaml
Comment (awojasinski-nordicsemi) [reply dts/bindings/dma/nordic,easydma-device.yaml]: moved `dts/bindings/dma/nordic,easydma-device.yaml` -> `dts/bindings/reserved-memory/memory-region.yaml`

**Review by carlocaione — DISMISSED** [review]

**Comment by awojasinski-nordicsemi** [comment]
> Can you also open an enhancement request to transition to use of memory-region property.

@galak is that ok for you? https://github.com/zephyrproject-rtos/zephyr/issues/46048

**Comment by galak** [comment]
Looks good, but nit pick, can you add the `memory-region-names` property as well from https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/reserved-memory/memory-region.yaml

**Comment by awojasinski-nordicsemi** [comment]
@galak now I noticed that in Linux memory-region is a phandle array type. I don't see any particular reason to change phandle to phandle-array in this PR. What are your thoughts about that?

File: dts/bindings/reserved-memory/memory-region.yaml
Comment (awojasinski-nordicsemi) [reply dts/bindings/reserved-memory/memory-region.yaml]: done

**Review by mbolivar-nordic — DISMISSED** [review]
blindly re-approving

**Review by carlocaione — DISMISSED** [review]

**Comment by awojasinski-nordicsemi** [comment]
force-pushed some changes to fix build errors regarding `phandle-array` property type

**Review by mbolivar-nordic — APPROVED** [review]
blindly re-approving again

**Review by carlocaione — APPROVED** [review]
```
while (1) {
    ACK;
}
```

**Review by galak — APPROVED** [review]

**Review by anangl — APPROVED** [review]

File: drivers/spi/spi_nrfx_spim.c
Comment (emob-nordic) [inline drivers/spi/spi_nrfx_spim.c:556]: Would this make any difference for DMA if this were aligned 4? @anangl

**Pull request merged**
